File: src/cli/mergeVeiculo.ts
import fs from "node:fs";
import path from "node:path";
import crypto from "node:crypto";
import { REPO_ROOT } from "../lib/repoRoot.js";
import { syncFipeNovoVeiculo } from "./atualizarFipeVeiculos.js";

const DBV = path.join(REPO_ROOT, "database", "veiculos.json");
const DBP = path.join(REPO_ROOT, "database", "parceiros.json");
const DBL = path.join(REPO_ROOT, "database", "parceiro-veiculo.json");

type Veiculo = Record<string, unknown> & { id?: string; placa?: string };
type Parceiro = { id: string; nome: string };
type Vinculo = { id: string; veiculoId: string; parceiroId: string };

export async function main(argv: string[]): Promise<void> {
  const novoPath = path.resolve(argv[0]!);
  const dono = argv[1]!.trim();
  const novo = JSON.parse(fs.readFileSync(novoPath, "utf8")) as Veiculo;

  const veic = JSON.parse(fs.readFileSync(DBV, "utf8")) as {
    veiculos: Veiculo[];
    atualizadoEm?: string;
  };
  const parc = JSON.parse(fs.readFileSync(DBP, "utf8")) as {
    parceiros: Parceiro[];
    atualizadoEm?: string;
  };
  const link = JSON.parse(fs.readFileSync(DBL, "utf8")) as {
    vinculos: Vinculo[];
    atualizadoEm?: string;
  };

  const placaN = String(novo.placa || "").toUpperCase();
  const ex = veic.veiculos.find(
    (v) => String(v.placa || "").toUpperCase() === placaN,
  );
  let acao: "atualizado" | "cadastrado";
  if (ex) {
    novo.id = ex.id;
    veic.veiculos = veic.veiculos.map((v) => (v === ex ? novo : v));
    acao = "atualizado";
  } else {
    novo.id = crypto.randomUUID();
    veic.veiculos.push(novo);
    acao = "cadastrado";
  }

  let p = parc.parceiros.find((x) => x.nome.toLowerCase() === dono.toLowerCase());
  if (!p) {
    p = { id: crypto.randomUUID(), nome: dono };
    parc.parceiros.push(p);
  }

  const vid = String(novo.id);
  const pid = p.id;
  link.vinculos = link.vinculos.filter((l) => l.veiculoId !== vid);
  link.vinculos.push({
    id: crypto.randomUUID(),
    veiculoId: vid,
    parceiroId: pid,
  });

  const today = new Date().toISOString().slice(0, 10);
  veic.atualizadoEm = today;
  parc.atualizadoEm = today;
  link.atualizadoEm = today;
  fs.writeFileSync(DBV, JSON.stringify(veic, null, 2), "utf8");
  fs.writeFileSync(DBP, JSON.stringify(parc, null, 2), "utf8");
  fs.writeFileSync(DBL, JSON.stringify(link, null, 2), "utf8");
  console.log(
    `Veiculo ${acao}: ${novo.placa} (id ${novo.id}) -> proprietario ${p.nome}`,
  );

  if (acao === "cadastrado") {
    await syncFipeNovoVeiculo(String(novo.placa || ""));
  }
}

File: .cursor/skills/cadastrar-veiculo/scripts/merge_veiculo.py
# -*- coding: utf-8 -*-
"""Merge veículo (sem id) em veiculos.json + parceiros + parceiro-veiculo."""
import json
import shutil
import subprocess
import sys
import uuid
from datetime import date
from pathlib import Path

ROOT = Path(__file__).resolve().parents[4]
DBV = ROOT / "database" / "veiculos.json"
DBP = ROOT / "database" / "parceiros.json"
DBL = ROOT / "database" / "parceiro-veiculo.json"
FIPE_SYNC = (
    ROOT
    / ".cursor"
    / "skills"
    / "cadastrar-veiculo"
    / "scripts"
    / "atualizar_fipe_veiculos.mjs"
)


def _sync_fipe_novo_veiculo(placa: str) -> None:
    """Atualiza fipe/fipeCodigo/fipeModelo/fipeValor/fipeReferencia na API FIPE."""
    if not placa or not FIPE_SYNC.is_file():
        if placa and not FIPE_SYNC.is_file():
            print(f"[aviso] Script FIPE ausente: {FIPE_SYNC}")
        return
    node = shutil.which("node")
    if not node:
        print(
            "[aviso] node nao encontrado no PATH; sincronize FIPE manualmente:\n"
            f'  node "{FIPE_SYNC}" --placa {placa}'
        )
        return
    try:
        r = subprocess.run(
            [node, str(FIPE_SYNC), "--placa", placa],
            cwd=str(ROOT),
            capture_output=True,
            text=True,
            encoding="utf-8",
            errors="replace",
            timeout=180,
        )
        out = (r.stdout or "").strip()
        err = (r.stderr or "").strip()
        if out:
            print(out)
        if r.returncode != 0:
            print(f"[aviso] FIPE sync exit {r.returncode}")
            if err:
                print(err)
        else:
            print("[fipe] campos FIPE atualizados na API")
    except subprocess.TimeoutExpired:
        print("[aviso] FIPE sync excedeu 180s; rode manualmente com --placa")
    except OSError as e:
        print(f"[aviso] FIPE sync: {e}")


def main():
    novo_path = Path(sys.argv[1])
    dono = sys.argv[2].strip()
    novo = json.loads(novo_path.read_text(encoding="utf-8"))

    veic = json.loads(DBV.read_text(encoding="utf-8"))
    parc = json.loads(DBP.read_text(encoding="utf-8"))
    link = json.loads(DBL.read_text(encoding="utf-8"))

    ex = next(
        (
            v
            for v in veic["veiculos"]
            if (v.get("placa") or "").upper() == (novo.get("placa") or "").upper()
        ),
        None,
    )
    if ex:
        novo["id"] = ex["id"]
        veic["veiculos"] = [novo if v is ex else v for v in veic["veiculos"]]
        acao = "atualizado"
    else:
        novo["id"] = str(uuid.uuid4())
        veic["veiculos"].append(novo)
        acao = "cadastrado"

    p = next((x for x in parc["parceiros"] if x["nome"].lower() == dono.lower()), None)
    if not p:
        p = {"id": str(uuid.uuid4()), "nome": dono}
        parc["parceiros"].append(p)

    vid = novo["id"]
    pid = p["id"]
    link["vinculos"] = [l for l in link["vinculos"] if l["veiculoId"] != vid]
    link["vinculos"].append(
        {"id": str(uuid.uuid4()), "veiculoId": vid, "parceiroId": pid}
    )

    today = date.today().isoformat()
    for d, path in [(veic, DBV), (parc, DBP), (link, DBL)]:
        d["atualizadoEm"] = today
        path.write_text(json.dumps(d, ensure_ascii=False, indent=2), encoding="utf-8")
    print(f"Veiculo {acao}: {novo['placa']} (id {novo['id']}) -> proprietario {p['nome']}")

    if acao == "cadastrado":
        _sync_fipe_novo_veiculo(novo.get("placa") or "")


if __name__ == "__main__":
    main()
